src/common/ad_tx.c

```diff
@@ -8,8 +8,152 @@
 #include <daos_srv/ad_mem.h>
 #include "ad_mem.h"
 
+struct umem_wal_tx_item {
+	d_list_t		ti_link;
+	struct umem_wal_tx	ti_utx;
+};
+
+struct ad_tls_cache {
+	int		atc_act_nr;
+	int		atc_tx_nr;
+	d_list_t	atc_act_list;
+	d_list_t	atc_tx_list;
+};
+
 static __thread struct ad_tx	*tls_tx;
 
+#define AD_TLS_CACHE_ENABLED	(1)
+#define TLS_ACT_NUM		(64)
+#define TLS_ACT_MAX		(512)
+#define TLS_TX_NUM		(16)
+
+static __thread struct ad_tls_cache tls_cache;
+
+static struct ad_act *
+tls_act_get(int opc, size_t size)
+{
+	struct ad_act	*act;
+
+#if AD_TLS_CACHE_ENABLED
+	if (opc != UMEM_ACT_COPY && tls_cache.atc_act_nr > 0) {
+		act = d_list_pop_entry(&tls_cache.atc_act_list, struct ad_act, it_link);
+		act->it_act.ac_opc = opc;
+		tls_cache.atc_act_nr--;
+		return act;
```

Comment:
How do you know the cached act size is large enough? I think we need to define a size used for cache, and when requested size is larger than cache size, allocate in-the-fly?

Reply:
for non-UMEM_ACT_COPY type action, all are with same size, right?
now it don't cache UMEM_ACT_COPY act.

Reply:
oh, I missed the (opc != UMEM_ACT_COPY) check. I think ACT_COPY are the major actions, what about cache larger act to satisfy most cases?

Reply:
OK, I just added the reusing for ACT_COPY, with max payload size of 512 that is enough for most cases. seems show with better vos_perf number.